server/src/main/java/org/cloudfoundry/identity/uaa/provider/oauth/ExternalOAuthProviderConfigurator.java

```diff
@@ -179,6 +181,30 @@ public List<IdentityProvider> retrieveActive(String zoneId) {
         return retrieveAll(true, zoneId);
     }
 
+    @Override
+    public List<IdentityProvider> retrieveActiveByType(final String type, final String zoneId) {
```

Comment:
this method is never call in /login but only retrieveAll

Reply:
I added this method to `ExternalOAuthProviderConfigurator` because it needs to correctly implement the interface (where the new method was added).

`retrieveActiveByType` has two differences to the method `retrieveAll` in this class:

- `retrieveAll` allows passing a parameter `activeOnly`; `retrieveActiveByType` does not allow querying for inactive IdPs
- `retrieveAll` returns the IdPs of type `oidc1.0` **and** `oauth2.0`; `retrieveActiveByType` only allows filtering for one type

I could address these two points by ...

1. changing the method to `retrieveByType` and adding a similar `activeOnly` parameter and
2. allowing multiple types to be passed to the method.

What do you think?

Reply:
This sounds reasonable, 

- retrieveAll returns the IdPs of type oidc1.0 and oauth2.0; retrieveActiveByType only allows filtering for one type

because the retrieveAll is used in realality and retrieveActiveByType is not used but the filter might be ok. Maybe if type is null return simply 
`oidc1.0 and oauth2.0`

Reply:
fixed with commit 7eefd26